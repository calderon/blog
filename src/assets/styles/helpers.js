import { css } from "styled-components"

const srOnly = css`
  position: absolute !important;
  width: 1px !important;
  height: 1px !important;
  padding: 0 !important;
  margin: -1px !important;
  overflow: hidden !important;
  clip: rect(0, 0, 0, 0) !important;
  white-space: nowrap !important;
  border: 0 !important;
`

const srOnlyFocusable = css`
  &:not(:focus) {
    ${srOnly};
  }
`

const skipLink = css`
  ${srOnly};
  ${srOnlyFocusable};
`

const _breakpoints = {
  xs: 480,
  sm: 640,
  md: 768,
  lg: 840,
  xl: 1024,
}

const devices = {
  xs: `(min-width: ${_breakpoints.xs}px)`,
  sm: `(min-width: ${_breakpoints.sm}px)`,
  md: `(min-width: ${_breakpoints.md}px)`,
  lg: `(min-width: ${_breakpoints.lg}px)`,
  xl: `(min-width: ${_breakpoints.xl}px)`,
}

export { srOnly, srOnlyFocusable, skipLink, devices }
