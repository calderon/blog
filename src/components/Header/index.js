import React, { useState, useLayoutEffect } from "react"
import styled from "styled-components"
import { rgba } from "polished"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import {
  faBars,
  faTimesCircle,
  faNewspaper,
  faAddressCard,
  faEnvelope,
} from "@fortawesome/free-solid-svg-icons"

import { devices } from "../../assets/styles/helpers"

import Heading from "../Heading"
import { Link, AnchorLink, AnchorLinkText } from "../Link"
import List from "../List"

const StyledHeader = styled.header.attrs(props => ({
  className: "header",
}))`
  grid-column-start: 1;
  grid-column-end: 4;
  background-color: ${props => props.theme.colors.bright};
  padding: 0 2.5rem;
  height: 7.5rem;
  display: flex;
  align-items: center;
  position: sticky;
  top: 0;

  @media ${devices.lg} {
    padding: 0 5rem;
  }

  &::before,
  &::after {
    content: "";
    height: 0.1rem;
    background: rgba(0, 0, 0, 0.1);
    position: absolute;
    top: 100%;
  }

  &::before {
    left: 50%;
    right: 50%;
    transition: left 300ms;
  }

  &::after {
    left: 50%;
    right: 50%;
    transition: right 300ms;
  }

  &.header--scrolled {
    &::before {
      left: 0;
    }

    &::after {
      right: 0;
    }
  }
`

const HeaderTitle = styled(Heading)`
  font-size: 3.6rem;
  font-weight: 700;
  letter-spacing: 0.13rem;
  line-height: 1.5;
  text-transform: lowercase;
  margin: 0;
  flex: 1;

  a {
    color: inherit;
    text-decoration: none;
  }
`

const HeaderMenu = styled(List)``

const HeaderMenuItem = styled.li`
  font-size: 3rem;
  text-transform: lowercase;
  margin-bottom: 1.25rem;

  &:last-child {
    margin-bottom: 0;
  }

  @media ${devices.lg} {
    font-size: 1.6rem;
    margin-right: 2.5rem;
    display: inline;
    margin-bottom: 0;
  }

  .header__navigation__link__icon {
    display: none;

    @media ${devices.lg} {
      display: inline-block;
      font-size: 2.4rem;
      vertical-align: middle;
      margin-right: 1rem;
    }
  }
`

const HeaderNavigation = styled.nav`
  &:not(.header__navigation--menuVisible) {
    .header__navigation__anchor--hideMenu {
      display: none;
    }
  }

  &.header__navigation--menuVisible {
    .header__navigation__anchor--showMenu {
      display: none;
    }

    ${HeaderMenu} {
      display: block;
      text-align: center;
      background: ${props => rgba(props.theme.colors.bright, 0.95)};
      position: fixed;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      padding: 7.5rem 5rem 5rem 5rem;
    }
  }

  ${HeaderMenu} {
    display: none;

    @media ${devices.lg} {
      display: block;
    }
  }

  ${AnchorLink} {
    font-size: 2.4rem;
    color: ${props => props.theme.palette.primary};

    &:hover {
      color: ${props => props.theme.palette.quinary};
    }
  }

  .header__navigation__anchor--hideMenu {
    position: relative;
  }
`

const Header = props => {
  const [scrolled, setScrolled] = useState(false)

  useLayoutEffect(() => {
    const handleScroll = e => {
      setScrolled(window.scrollY > 0)
    }

    window.addEventListener("scroll", handleScroll)

    return () => {
      window.removeEventListener("scroll", handleScroll)
    }
  }, [])

  const handleShowMenu = e => {
    e.preventDefault()

    document
      .querySelector(".header__navigation")
      .classList.add("header__navigation--menuVisible")

    document.body.classList.add("unscrolled")
  }

  const handleCloseMenu = e => {
    e.preventDefault()

    document
      .querySelector(".header__navigation")
      .classList.remove("header__navigation--menuVisible")

    document.body.classList.remove("unscrolled")
  }

  return (
    <StyledHeader className={scrolled ? "header--scrolled" : ""}>
      <HeaderTitle className="header__title">
        <Link className="header__title__link" to="/">
          Daniel Calder√≥n
        </Link>
      </HeaderTitle>

      <HeaderNavigation className="header__navigation">
        <AnchorLink
          to="#headerMenu"
          className="header__navigation__anchor header__navigation__anchor--showMenu"
          hiddenFrom="lg"
          onClick={handleShowMenu}
        >
          <AnchorLinkText>Go to navigation menu</AnchorLinkText>
          <FontAwesomeIcon icon={faBars} />
        </AnchorLink>

        <Heading className="header__navigation_title" srOnly>
          Navigation menu
        </Heading>

        <HeaderMenu
          id="headerMenu"
          className="header__navigation__list"
          unstyled
        >
          <HeaderMenuItem>
            <Link
              className="header__navigation__link header__navigation__link--blog"
              to="/blog"
            >
              <FontAwesomeIcon
                icon={faNewspaper}
                className="header__navigation__link__icon"
              />
              Blog
            </Link>
          </HeaderMenuItem>
          <HeaderMenuItem>
            <Link
              className="header__navigation__link header__navigation__link--about"
              to="/about"
            >
              <FontAwesomeIcon
                icon={faAddressCard}
                className="header__navigation__link__icon"
              />
              About me
            </Link>
          </HeaderMenuItem>
          <HeaderMenuItem>
            <Link
              className="header__navigation__link header__navigation__link--contact"
              to="#contact"
            >
              <FontAwesomeIcon
                icon={faEnvelope}
                className="header__navigation__link__icon"
              />
              contact
            </Link>
          </HeaderMenuItem>
        </HeaderMenu>

        <AnchorLink
          to="#headerMenu"
          className="header__navigation__anchor header__navigation__anchor--hideMenu"
          hiddenFrom="lg"
          onClick={handleCloseMenu}
        >
          <AnchorLinkText>Go up</AnchorLinkText>
          <FontAwesomeIcon icon={faTimesCircle} />
        </AnchorLink>
      </HeaderNavigation>
    </StyledHeader>
  )
}

export default Header
