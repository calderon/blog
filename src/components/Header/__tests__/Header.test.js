import React from "react"
import { render } from "../../../../__tests__/helpers/render"

import Header from "../"

const renderComponent = ({ ...props } = {}) => render(<Header {...props} />)

describe("Header", () => {
  it("renders correctly", () => {
    const header = renderComponent()
    expect(header).toMatchSnapshot()
  })
})
