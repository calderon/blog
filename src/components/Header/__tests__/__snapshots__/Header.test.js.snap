// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="sc-AxiKw eSCqxD header"
      >
        <h1
          class="sc-AxjAm sc-AxhCb gdZBxN header__title heading heading--h1"
        >
          <a
            aria-current="page"
            class="header__title__link"
            href="/"
          >
            Daniel Calderón
          </a>
        </h1>
        <nav
          class="sc-AxhUy kMyssT header__navigation"
        >
          <h1
            class="sc-AxjAm dxUEld header__navigation_title heading heading--h1"
          >
            Navigation menu
          </h1>
          <ul
            class="sc-AxirZ sc-AxgMl loCSKX header__navigation__list list"
          >
            <li
              class="sc-AxheI cOzwHk"
            >
              <a
                aria-current="page"
                class="header__navigation__link header__navigation__link--blog"
                href="/"
              >
                Blog
              </a>
            </li>
            <li
              class="sc-AxheI cOzwHk"
            >
              <a
                class="header__navigation__link header__navigation__link--about"
                href="/about"
              >
                About me
              </a>
            </li>
          </ul>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="sc-AxiKw eSCqxD header"
    >
      <h1
        class="sc-AxjAm sc-AxhCb gdZBxN header__title heading heading--h1"
      >
        <a
          aria-current="page"
          class="header__title__link"
          href="/"
        >
          Daniel Calderón
        </a>
      </h1>
      <nav
        class="sc-AxhUy kMyssT header__navigation"
      >
        <h1
          class="sc-AxjAm dxUEld header__navigation_title heading heading--h1"
        >
          Navigation menu
        </h1>
        <ul
          class="sc-AxirZ sc-AxgMl loCSKX header__navigation__list list"
        >
          <li
            class="sc-AxheI cOzwHk"
          >
            <a
              aria-current="page"
              class="header__navigation__link header__navigation__link--blog"
              href="/"
            >
              Blog
            </a>
          </li>
          <li
            class="sc-AxheI cOzwHk"
          >
            <a
              class="header__navigation__link header__navigation__link--about"
              href="/about"
            >
              About me
            </a>
          </li>
        </ul>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
